### PR TITLE
Use RWLock for Format's CSS cache

### PR DESCRIPTION
**Pull request by gkaikoponen**
This should ensure read lock contention will never be a bottleneck for heavily concurrent code using Format.

### CODE REVIEW

**Comment by gkaikoponen** [comment]
Just realized this PR is no good as-is as the read code contains LRU manipulations

**Comment by gkaikoponen** [comment]
The correct way to approach this is to grab the read lock while checking only the most recent entry, then switch to the write lock if it does match.

**Comment by gkaikoponen** [comment]
@alecthomas FYI I [amended the base branch](https://github.com/gkaikoponen/chroma/commit/6ca44f097a9195f3d68c80733b519e40112d9163) here to show what this would look like with the write lock not grabbed while reading the freshest cache entry. Up to you if that's worth the complexity.

File: formatters/html/html.go [hunk 1]
```diff
@@ -583,25 +583,33 @@ func newStyleCache(f *Formatter) *styleCache {
...
+		// Look for an existing entry.
+		for i := len(l.cache) - 1; i >= 0; i-- {
+			entry := l.cache[i]
+			if entry.style == style {
+				// Top of the cache, no need to adjust the order.
+				if i == len(l.cache)-1 {
+					return entry.cache
+				}
+				// Move this entry to the end of the LRU
+				copy(l.cache[i:], l.cache[i+1:])
```
Comment (alecthomas) [inline formatters/html/html.go:599]: This mutates the cache so needs the write lock

**Comment by alecthomas** [comment]
Thanks for that. I think it's okay as-is for now, particularly as any lock contention will be minimal compared to the overall cost of the parent calls.

**Pull request closed without merging**